File: csharp/beer-song/BeerSong.cs
using System;

public static class BeerSong
{
    public static string Recite(int startBottles, int takeDown)
    {
        const string word = "bottle";
        var result = "";

        var current = startBottles;
        for (int i = 1; i <= takeDown; i++)
        {
            var next = current - 1;
            if (result != "")
            {
                result += "\n\n";
            }

            if (next >= 0)
            {
                result += $"{current} {GetWord(word, current)} of beer on the wall, {current} {GetWord(word, current)} of beer.";
            
                if (next == 0)
                {
                    result += "\nTake it down and pass it around, no more bottles of beer on the wall.";
                }
                else
                {
                    result += $"\nTake one down and pass it around, {next} {GetWord(word, next)} of beer on the wall.";
                }
            }
            else
            {
                result += "No more bottles of beer on the wall, no more bottles of beer.\n" +
                          "Go to the store and buy some more, 99 bottles of beer on the wall.";
            }
           

            current = next;
        }

        return result;
    }

    private static string GetWord(string word, int number) => number > 1 ? word + "s" : word;
}

File: cpp/beer-song/beer_song.cpp
#include "beer_song.h"

namespace beer_song
{

    std::string verse(int n)
    {
        int next = n - 1;

        if (next >= 0)
        {
            std::string word = n > 1 ? "bottles" : "bottle";
            std::string result = std::to_string(n) + " " + word + " of beer on the wall, " + std::to_string(n) + " " + word + " of beer.\n";

            if (next == 0)
            {
                result += "Take it down and pass it around, no more bottles of beer on the wall.\n";
            }
            else
            {
                std::string next_word = next > 1 ? "bottles" : "bottle";
                result += "Take one down and pass it around, " + std::to_string(next) + " " + next_word + " of beer on the wall.\n";
            }

            return result;
        }

        return "No more bottles of beer on the wall, no more bottles of beer.\nGo to the store and buy some more, 99 bottles of beer on the wall.\n";
    }

    std::string sing(int from, int to)
    {
        std::string result;
        for (int i = from; i >= to; i--)
        {
            std::string v = verse(i);

            if (!result.empty())
            {
                result += "\n";
            }

            result += v;
        }

        return result;
    }

    std::string sing(int from)
    {
        return sing(from, 0);
    }

} // namespace beer_song
